lib/Sema/MiscDiagnostics.cpp

```diff
@@ -2645,6 +2645,44 @@ class OpaqueUnderlyingTypeChecker : public ASTWalker {
     // If there are no candidates, then the body has no return statements, and
     // we have nothing to infer the underlying type from.
     if (Candidates.empty()) {
+      // We try to find if the last element of the `Body` multi element
+      // `BraceStmt` is an expression that produces a value that satisfies all
+      // the opaque type requirements and if that is the case, it means we can
+      // suggest a fix-it note to add an explicit `return`.
+      if (Body->getNumElements() > 1) {
+        auto element = Body->getLastElement();
+        // Let's see if the last statement would make for a valid return value.
+        if (auto expr = element.dyn_cast<Expr *>()) {
+          bool conforms = true;
+
+          for (auto requirement :
+               OpaqueDecl->getOpaqueInterfaceGenericSignature().getRequirements()) {
+            auto protocolConformance =
+                TypeChecker::conformsToProtocol(expr->getType()->getRValueType(),
+                                                requirement.getProtocolDecl(),
+                                                Implementation->getModuleContext(),
+                                                /*allowMissing=*/ false);
+            if (protocolConformance.isInvalid()) {
+              conforms = false;
+              break;
+            }
+          }
+
+          // If all requirements are fulfilled, we offer to insert `return` to
+          // fix the issue.
+          if (conforms) {
+            Implementation->diagnose(diag::opaque_type_no_underlying_type_candidates);
+            Ctx.Diags
+                .diagnose(expr->getStartLoc(),
+                          diag::opaque_type_missing_return_last_expr_note)
+                .fixItInsert(expr->getStartLoc(), "return ");
+            return;
+          }
+        }
+      }
+
+      // We have not found any candidates for where adding a `return` would fix
+      // the issue and are therefore falling back to the generic error message.
       Implementation->diagnose(diag::opaque_type_no_underlying_type_candidates);
```

Comment:
Also, maybe we could move this logic here... since diagnostic will always be emitted we can put this logic after it and if conditions are met we emit only the fix-it :)